java-frontend/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -845,6 +874,35 @@ private void buildNewClass(NewClassTree tree) {
     build(Lists.reverse(tree.arguments()));
   }
 
+  private void handleExceptionalPaths(Symbol symbol) {
+    TryStatement pop = enclosingTry.pop();
+    TryStatement tryStatement;
+    if(!enclosedByCatch.peek()) {
+      tryStatement = pop;
+    } else {
+      tryStatement = enclosingTry.peek();
+    }
+    enclosingTry.push(pop);
+    if(pop != outerTry) {
+      currentBlock = createBlock(currentBlock);
+      currentBlock.exceptions.add(exitBlocks.peek());
+    }
+    if (symbol.isMethodSymbol()) {
+      List<Type> thrownTypes = ((Symbol.MethodSymbol) symbol).thrownTypes();
+      if (!thrownTypes.isEmpty()) {
+        thrownTypes.forEach(thrownType -> {
+          for (Type caughtType : tryStatement.catches.keySet()) {
```

Comment:
I would store the `keySet` in a variable before calling `thrownTypes.foreach` in order to avoid recalculating it each time.